src/states/Preroll.js

```diff
@@ -0,0 +1,221 @@
+import videojs from 'video.js';
+
+import {AdState, ContentPlayback, BeforePreroll} from '../states.js';
+import cancelContentPlay from '../cancelContentPlay.js';
+import {startAdBreak, endAdBreak} from '../adBreak.js';
+
+/*
+ * This state encapsulates waiting for prerolls, preroll playback, and
+ * content restoration after a preroll.
+ */
+export default class Preroll extends AdState {
+
+  init(player, adsReady) {
+    // Loading spinner from now until ad start or end of ad break.
+    player.addClass('vjs-ad-loading');
+
+    // Start the clock ticking for ad timeout
+    this._timeout = player.setTimeout(function() {
+      player.trigger('adtimeout');
+    }, player.ads.settings.prerollTimeout);
+
+    // If adsready already happened, lets get started. Otherwise,
+    // wait until onAdsReady.
+    if (adsReady) {
+      this.handleAdsReady();
+    } else {
+      this.adsReady = false;
+    }
+  }
+
+  onAdsReady(player) {
+    if (!player.ads.inAdBreak() && !player.ads.isContentResuming()) {
+      player.ads.debug('Received adsready event (Preroll)');
+      this.handleAdsReady();
+    } else {
+      videojs.log.warn('Unexpected adsready event (Preroll)');
+    }
+  }
+
+  /*
+   * Ad integration is ready. Let's get started on this preroll.
+   */
+  handleAdsReady() {
+    this.adsReady = true;
+    if (this.player.ads.nopreroll_) {
+      this.noPreroll();
+    } else {
+      this.readyForPreroll();
+    }
+  }
+
+  /*
+   * If there is no preroll, play content instead.
+   */
+  noPreroll() {
+    this.player.ads.debug('Skipping prerolls due to nopreroll event');
+    this.transitionTo(ContentPlayback);
+  }
+
+  /*
+   * Fire the readyforpreroll event. If loadstart hasn't happened yet,
+   * wait until loadstart first.
+   */
+  readyForPreroll() {
+    const player = this.player;
+
+    // Signal to ad plugin that it's their opportunity to play a preroll.
+    if (player.ads._hasThereBeenALoadStartDuringPlayerLife) {
+      player.ads.debug('Triggered readyforpreroll event (Preroll)');
+      player.trigger('readyforpreroll');
+
+    // Don't play preroll before loadstart, otherwise the content loadstart event
+    // will get misconstrued as an ad loadstart. This is only a concern for the
+    // initial source; for source changes the whole ad process is kicked off by
+    // loadstart so it has to have happened already.
+    } else {
+      player.one('loadstart', () => {
+        player.ads.debug('Triggered readyforpreroll event (loadstart)');
+        player.trigger('readyforpreroll');
+      });
+    }
+  }
+
+  /*
+   * Don't allow the content to start playing while we're dealing with ads.
+   */
+  onPlay(player) {
+    player.ads.debug('Received play event (Preroll)');
+
+    if (!this.inAdBreak() && !this.isContentResuming()) {
+      cancelContentPlay(this.player);
+    }
+  }
+
+  /*
+   * adscanceled cancels all ads for the source. Play content now.
+   */
+  onAdsCanceled(player) {
```

Comment:
this seems like a value that should be passed along to subsequent ad breaks

Reply:
I'm on board with removing it

Reply:
Cool. Again, to limit the scope of this PR, I'd like to do that in a separate PR.